[fix] [broker] The broker has two identical Persitenttopics

poorbarcode opened the pull request:
### Motivation

#### Problem occur

With the transaction feature, we send and receive messages, and at the same time, execute `admin API: unload namespace` 1000 times. Then the problem occur: Consumer could not receive any message, and there has no error log. After that we tried `admin API: get topic stats`, and the response showed only producers are registered on topic, and no consumers are registered on topic, but consumer stat is `Ready` in the client. This means that the state of the consumer is inconsistent between the broker and the client.

#### Location problem

Then we found the problem: Two PersistentTopic which have the same name registered at a broker node, consumer registered on one (aka `topic-c`), and producer registered on another one (aka `topic-p`). At this time, when we send messages, the data flow like this : 

```text
client: producer sends a message

broker: handle cmd-send

broker: find the topic by name, it is "topic-p"

broker: find all subscriptions registered on "topic-p"

broker: found one subscription, but it has no consumers registered

broker: no need to send the message to the client
``` 

But the consumer exactly registered on another topic: `topic-c`, so consumer could not receive any message.

#### Repreduce

> *How to reproduce two topics registered at the same broker node ?*

Make  `transaction buffer recover`, `admin unload namespace`, `client create consumer`, ` client create producer` executed at the same time, the process flow like this (at the step-11, the problem occurs ):

| Time | `transaction buffer recoverr` | `admin unload namespace` | `client create consumer` | `client create producer` |
| ----------- | ----------- | ----------- | ----------- | ----------- |
| 1 | TB recover |  |  |  |
| 2 | TB recover failure | topic.unload |  |  |
| 3 | topic.close(false) | topic.close(true) |  |  |
| 4 | brokerService.topics.remove(topicName) |  |  |  |
| 5 | remove topic finish |  | lookup |  |
| 6 |  |  | create `topic-c` |  |
| 7 |  |  | consumer registered on `topic-c` |  |
| 8 |  | brokerService.topics.remove(topic) |  |  |
| 9 |  | remove `topic-c` finish |  | lookup |
| 10 |  |  |  | create `topic-p`  |
| 11 |  |  |  | producer registered on `topic-p`  |

- Each column means the individual process. e.g. `client create consumer`,  `client create producer`.
- Multiple processes are going on at the same time, and all effet the `brokerService.topics`.
- Column `Time` is used only to indicate the order of each step, not the actual time.
- The important steps are explained below:

> step 3

Even if persistent topic property`isClosingOrDeleting` have already changed to `true`, it still can be executed another once, see line-1247::

https://github.com/apache/pulsar/blob/f230d15ffcd5f74cca13bd23b35ace784d6f8ce6/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L1240-L1249 

Whether close can be executed depends on two predicates: `is closing` or `@param closeWithoutWaitingClientDisconnect is true`. This means that method `topic.close` can be reentrant executed when `@param closeWithoutWaitingClientDisconnect` is true, and in the implementation of `admin API: unload namespace` the parameter `closeWithoutWaitingClientDisconnect` is exactly `true`.

https://github.com/apache/pulsar/blob/f230d15ffcd5f74cca13bd23b35ace784d6f8ce6/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java#L723-L725

So when `transaction buffer recover fail` and `admin unload namespace` is executed at the same time, and `transaction buffer recover fail` before `admin unload namespace`, the topic will be removed from `brokerService.topics` twice.

> step-4 / step-8

Because of the current implementation of `BrokerService. removeTopicFromCache` use cmd `map.remove(key)`, not use `map.remove(key, value)`, So this cmd can remove any value in the map, even if it's not the desired one.

https://github.com/apache/pulsar/blob/f230d15ffcd5f74cca13bd23b35ace784d6f8ce6/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java#L1956

To sum up: We should make these two changes: 

- Make method `topic.close` non-reentrant. Also prevent reentrant between `topic.close` and `topic.delete`.
- Use cmd  `map.remove(key, value)` instead of `map.remove(key)` in implementation of `BrokerService. removeTopicFromCache`. This change will apply to both scenes: `topic.close` and `topic.delete`.

### Other Modifications

In the current implementation, if closing the ledger fails, it determines that the closing topic failed. Then will reset the topic stat to `no-fenced`. But it changes two states [`isFenced`, `isClosingOrDeleting`] without locking,  this could not guarantee consistency between them.  I will fix it in this PR too (this change may not be relevant to current subject).

https://github.com/apache/pulsar/blob/f230d15ffcd5f74cca13bd23b35ace784d6f8ce6/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L1307-L1312

### PR Relations

#16240

When repeating creates the same topic at one broker node, the following phenomena occur: Transaction pending ack store reuses the cached managed cursor object when initializing constructor, the process flow like this: 

| Time | `client create consumer 1` | `client create consumer 2` |
| -----------  | ----------- | ----------- |
| 1 | lookup | lookup |
| 1 | create `topic'` |  |
| 2 | create `subscription'` | create `topic''` |
| 2 | open new `pending_ackledger` | create `subscription''` | 
| 3 | open new `pending_ack_cursor` | reuse the cached `pending_ackledger` | 
| 4 |   | reuse the cached `pending_ack_cursor`|

If the Transaction pending ack store reuses the managed cursor in the cache, it will cause the task `replay pending ack log` to loop forever. This PR solves the problem of “Repeat creates topic”, and also eliminates the possibility of reuse pending-ack-cursor: #16240.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs, and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

poorbarcode commented:
@gaoran10 @congbobo184 @Technoboy-  Please take a look. ^_^

poorbarcode commented:
/pulsarbot run-failure-checks

poorbarcode commented:
/pulsarbot run-failure-checks

poorbarcode commented:
Hi @codelipenghui 

> We also should consider to avoid a topic can be closed more than once.

I have appended the reason why `topic.close` was executed twice to the Motivation, and in this PR I've overwritten the `topic.close` to fix it. I also added a lock to the 'reset topic stat to UN-fenced' operation, could you review the code.

 > We can use map.compute() to simplify the logic? 

Unfortunately, we can't use `map.compute` tto simplify the logic.

> And looks like we don't need to wait for the future complete, because we already have the topic reference here.
 Yes, I have fixed it. The current implementation doesn't need to wait for the future to complete.

I also rewritten the Motivation of this PR to make it easier to understand. Thanks.

poorbarcode commented:
@eolivelli @lhotari @michaeljmarshall @gaozhangmin @Jason918 @nicoloboschi   Please take a look, if you have time. Thanks.

poorbarcode commented:
/pulsarbot run-failure-checks

poorbarcode commented:
@merlimat Could you take a look?

poorbarcode commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:
```diff
@@ -1918,15 +1918,38 @@ public AuthorizationService getAuthorizationService() {
+            }
+            // If topic is not in Cache, do nothing.
+            return CompletableFuture.completedFuture(null);
+        });
```
> We can use map.compute() to simplify the logic?  And looks like we don't need to wait for the future complete

Yes， I've rewritten the logic.

poorbarcode replied:
> We also should consider to avoid a topic can be closed more than once.

I have appended the reason why `topic.close` was executed twice to the Motivation, and in this PR I've overwritten the `topic.close` to fix it. I also added a lock to the 'reset topic stat to UN-fenced' operation, could you review the code.

poorbarcode commented:
/pulsarbot rerun-failure-checks

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:
```diff
@@ -1953,15 +1953,46 @@ public AuthorizationService getAuthorizationService() {
+            return CompletableFuture.completedFuture(null);
+        }
+        // We don't need to wait for the future complete, because we already have the topic reference here.
+        if (!createTopicFuture.isDone()){
```
we can save us from using getNow and we can chain the CompletableFuture with "thenCompose"

poorbarcode replied:
> I think that we should wait for this future to complete, otherwise it may create another Topic

One comment suggestion here: "don't need to wait for the future complete" 
https://github.com/apache/pulsar/pull/16247#issuecomment-1170101266

> we can save us from using getNow and we can chain the CompletableFuture with "thenCompose"

Already use "thenCompose" instead "getNow". Thanks

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -1243,86 +1247,146 @@ public CompletableFuture<Void> close() {
+                    return this.fullyCloseFuture;
+                } else {
+                    // Why not return this.fullyCloseFuture ?
+                    // I don't know, just keep the same implementation as before.
```
please remove this kind of comments "I don't know"

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -1243,86 +1247,146 @@ public CompletableFuture<Void> close() {
+        // Close limiters.
+        try {
+            closeLimiters();
+        } catch (Throwable t){
```
why are we catching "Throwable" ? this is usually a bad practice

poorbarcode replied:
Already instead "Throwable" to "Exception". Thanks

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -1243,86 +1247,146 @@ public CompletableFuture<Void> close() {
+            closePhase2Future = closeClientsFuture.thenCompose(__ -> asyncCloseLedger());
+        }
+        // Complete resultFuture. If managed ledger close failure, reset topic to resume.
+        closePhase2Future.thenApply(__ -> {
```
nit: we can use "whenComplete" instead of theApply/exceptionally

poorbarcode replied:
Already use "whenComplete" instead of "theApply". Thanks

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -1243,86 +1247,146 @@ public CompletableFuture<Void> close() {
+        }).exceptionally(exception -> {
+            log.error("[{}] Error closing topic", topic, exception);
+            // Restart rate-limiter after close managed ledger failure. Success is not guaranteed.
+            // TODO Guarantee rate-limiter open finish.
```
please to not leave "TODOs", open a new GH ticket and link it

poorbarcode replied:
Already remove "TODO". Thanks

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -1243,86 +1247,146 @@ public CompletableFuture<Void> close() {
+            // TODO Guarantee rate-limiter open finish.
+            try {
+                restartLimitersAfterCloseTopicFail();
+            } catch (Throwable t){
```
please do not catch Throwable blindly

poorbarcode replied:
Already instead "Throwable" to "Exception". Thanks

poorbarcode commented:
/pulsarbot rerun-failure-checks

poorbarcode commented:
/pulsarbot rerun-failure-checks

poorbarcode commented:
Hi @eolivelli 

Could you review this PR again?

Jason918 commented:
@codelipenghui @eolivelli @lhotari PTAL

congbobo184 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:
```diff
@@ -1953,15 +1953,50 @@ public AuthorizationService getAuthorizationService() {
+    }
+
+    public CompletableFuture<Void> removeTopicFromCache(String topic){
+        return removeTopicFromCache(topic, (CompletableFuture) null);
```
```suggestion
        return removeTopicFromCache(topic, (CompletableFuture<Optional<Topic>>) null);
```

poorbarcode replied:
Already fixed

congbobo184 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -1243,86 +1247,145 @@ public CompletableFuture<Void> close() {
+        // Close limiters.
+        try {
+            closeLimiters();
+        } catch (Exception t){
```
why should catch the Exception?

poorbarcode replied:
Just to keep the logic the same as before: `close limiters` failure does not affect `topic close`.

congbobo184 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -1243,86 +1247,145 @@ public CompletableFuture<Void> close() {
+        }
+
+        // Close client components.
+        CompletableFuture<Void> closeClientsFuture = asyncCloseClients();
```
```suggestion
        CompletableFuture<Void> closeClientsFuture = asyncCloseClients(boolean closeWithoutWaitingClientDisconnect);
```

poorbarcode replied:
Just to keep the logic the same as before: Restart rate-limiter after close managed ledger failure. Success is not guaranteed.

congbobo184 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -1243,86 +1247,145 @@ public CompletableFuture<Void> close() {
+                log.error("[{}] Error closing topic", topic, ex);
+                // Restart rate-limiter after close managed ledger failure. Success is not guaranteed.
+                try {
+                    restartLimitersAfterCloseTopicFail();
```
why should catch the Exception?

poorbarcode replied:
Just to keep the logic the same as before: Restart rate-limiter after close managed ledger failure. Success is not guaranteed.

poorbarcode commented:
To make this PR easy to Review, it was split into two other PR:

- Make method `topic.close` non-reentrant. Also prevent reentrant between `topic.close` and `topic.delete`.
  - fixed by PR #17524
- Use cmd  `map.remove(key, value)` instead of `map.remove(key)` in implementation of `BrokerService. 
  - fixed by PR #17526

poorbarcode commented:
To make this PR easy to Review, it was split into two other PR:

- Make method `topic.close` non-reentrant. Also prevent reentrant between `topic.close` and `topic.delete`.
  - fixed by PR #17524
- Use cmd  `map.remove(key, value)` instead of `map.remove(key)` in implementation of `BrokerService. 
  - fixed by PR #17526
